Connections pane integration with the variables pane

dfalbel opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://connect.posit.it/positron-wiki/pull-requests.html
* Ensure that the code is up-to-date with the `main` branch.
-->

### Intent

This PR integrates the variables pane with the connections pane. When variables are assigned and they are known to the connections pane we display a small icon, that triggers the registration of that connection with the connections pane.
Addresses https://github.com/posit-dev/positron/issues/2503


### Approach

> Describe the approach taken and the tradeoffs involved if non-obvious; add an overview of the solution if it's complicated.

1. Added a new inspector that catches supported connections and enables the `view` buttom for them:

https://github.com/posit-dev/positron/blob/f233f3ee41366f6f9110abc16310e3418f4b7300/extensions/positron-python/pythonFiles/positron/positron_ipykernel/inspectors.py#L902-L909

2. Made the variables pane preview handling depend on the type of object, and for supported values we will
    enable register the object with the connections service:

https://github.com/posit-dev/positron/blob/f233f3ee41366f6f9110abc16310e3418f4b7300/extensions/positron-python/pythonFiles/positron/positron_ipykernel/variables.py#L519-L537

 3. Add listeners to variable updates/deletions in the variables service allowing the connections service to handle those events, if required.

https://github.com/posit-dev/positron/blob/f233f3ee41366f6f9110abc16310e3418f4b7300/extensions/positron-python/pythonFiles/positron/positron_ipykernel/variables.py#L157-L169


### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues.

A small video below showing different scenarios that we expect this PR to work with:

https://github.com/posit-dev/positron/assets/4706822/a08050c9-b84d-4b35-a804-fe051f201b9d

dfalbel committed:
Initial work on variables pane support

dfalbel committed:
Rename method

dfalbel committed:
Somy pyright fixes

dfalbel committed:
Correctly create the tuple

dfalbel committed:
Add a connections inspector

dfalbel committed:
kind = "connection" support in the variables contract

dfalbel committed:
Customize icon depending on the variable type.

dfalbel committed:
Better handling of updates and deletions.

dfalbel committed:
Add unit tests for the variables pane integration.

dfalbel committed:
re-use make msg

dfalbel committed:
fixed paths

dfalbel committed:
Handle front-ent initiated comm_close messages.

isabelizimm reviewed (commented):
Some random teeny nits, but overall, this works really nicely for Python 3.10+! It looks like `sqlite3` Connections are not registering on Python 3.9 and lower. I took a look at the different versions [(this is the diff of 3.9 and 3.10)](https://github.com/python/cpython/compare/3.9...3.10), but wasn't quite sure where to look. There's not much actual Python code there, so it's probably somewhere on the C side, under `Modules/_sqlite` 😅 

One more random thing: when users click on the DB icon in the Variables pane, should it bring focus to the Connections pane ?

dfalbel committed:
Handle front-end initiated comm closes

dfalbel committed:
make it a set instead.

dfalbel committed:
upgrade ruff and reformat

dfalbel committed:
Specify type

dfalbel committed:
Correctly support older python versions.

dfalbel committed:
Add a comment on why we have this line

dfalbel committed:
Display the connections pane

The connections pane is now displayed when a
connection is openend from the variables pane,
or using `%connection_show` line magic.

isabelizimm reviewed (approved):
This looks good for the supported versions of Python! It also feels much faster than before ☄️ 

One small thing-- if you have a Connection open and then shut down the runtime (without disconnecting), you'll get this in the logs:

```
[Python]   File "/Users/isabelzimmerman/code/positron/extensions/positron-python/python_files/positron/positron_ipykernel/positron_ipkernel.py", line 374, in do_shutdown
[Python]     self.connections_service.shutdown()
[Python]   File "/Users/isabelzimmerman/code/positron/extensions/positron-python/python_files/positron/positron_ipykernel/connections.py", line 346, in shutdown
[Python]     for comm_id in self.comms.keys():
[Python] RuntimeError: dictionary changed size during iteration
```

I think if you change that line to `list(self.comms.keys())` it should be good!

dfalbel committed:
Fix shutdown error chaning the dict during the iteration.

jmcphers reviewed (changes requested):
Nice!

One thing that isn't clear to me is why we need the `createEventEmitter` bits. It looks like they are duplicating the codegen'ed event `onDidFocus`. If you want to expose that event in ConnectionsComm, you can do so just by creating a public-facing `Event` and then assigning it to the wrapped `onDidFocus` from the generated class.

dfalbel committed:
take care of disposables
